avr-hal-generic/src/i2c.rs

```diff
@@ -497,9 +592,27 @@ macro_rules! impl_twi_i2c {
                 let last = buffer.len() - 1;
                 for (i, byte) in buffer.iter_mut().enumerate() {
                     if i != last {
+                        // (datasheet)
+                        //
+                        //    Bit 6 – TWEA: TWI Enable Acknowledge Bit
+                        //
+                        //    The TWEA bit controls the generation of the acknowledge pulse. If the TWEA bit is written to one, the ACK
+                        //    pulse is generated on the TWI bus if the following conditions are met:
+                        //    1. The device’s own slave address has been received.
+                        //    2. A general call has been received, while the TWGCE bit in the TWAR is set.
+                        //    3. A data byte has been received in Master Receiver or Slave Receiver mode.
+                        //    By writing the TWEA bit to zero, the device can be virtually disconnected from the 2-wire Serial Bus temporarily.
+                        //    Address recognition can then be resumed by writing the TWEA bit to one again.
                         self.p.$twcr.write(|w| w.$twint().set_bit().$twen().set_bit().$twea().set_bit());
                         self.wait();
                     } else {
+                        // QUESTION: Why is this behaviour different for the last byte?
```

Comment:
Curious as to the bifurcation when writing the byte array. Why is `$twea` not set for the final byte? I thought the final byte required an ACK as well.

Reply:
The datasheet text you insterted below explains it:  After the last byte, master send NACK and then a stop condition.

Reply:
I see why I was confused now.

So we have read the last byte in `read_data()`, and we set TWINT and TWEN high:

https://github.com/Rahix/avr-hal/blob/9f9d6972e4ac7c149e2644bbc52c47cca81838f4/avr-hal-generic/src/i2c.rs#L397

Then call `wait()` which waits for TWINT to be cleared:

https://github.com/Rahix/avr-hal/blob/9f9d6972e4ac7c149e2644bbc52c47cca81838f4/avr-hal-generic/src/i2c.rs#L356

Then wherever `read_data()` was invoked from can invoke a STOP, like here in `blocking_read()`:

https://github.com/Rahix/avr-hal/blob/9f9d6972e4ac7c149e2644bbc52c47cca81838f4/avr-hal-generic/src/i2c.rs#L477

The `stop()` is here:

https://github.com/Rahix/avr-hal/blob/9f9d6972e4ac7c149e2644bbc52c47cca81838f4/avr-hal-generic/src/i2c.rs#L420

My question has evolved to:

1. (minor question) I don't see a `restart` method on the API... is that intentional?
1. The datasheet defines the STOP as being invoked exactly as in the `stop()` method shown. My question is why the intervening writes? Here's what I see:

a) write TWINT TWEN high
b) wait for TWINT to clear
c) write TWINT TWEN TWSTO high

Why do we need `a` and `b`?

Reply:
> a) write TWINT TWEN high

this clears the TWINT bit, as I described in the other comments

>  b) wait for TWINT to clear

actually, the line

https://github.com/Rahix/avr-hal/blob/9f9d6972e4ac7c149e2644bbc52c47cca81838f4/avr-hal-generic/src/i2c.rs#L357

waits for it to become set again ("while bit is cleared, loop").